Upgrade Welcome theme to PatternFly 5

jonkoops opened the pull request:
This PR upgrades the Welcome theme to PatternFly 5, as it was previously stuck on version 3. This PR also makes a couple of design and UX changes, namely:

- The user is now redirected to the Administration Console if the administrative user has already been created.
- The page is now responsive and will adapt the layout on devices with smaller screens.
- The background and Keycloak logo are now vector based (SVG).
- The favicon is now loaded from the 'common' resources, mitigating the need to keep another asset around just for the welcome page.
- The form to register a new user now has the appropriate `autocomplete` attributes, allowing browsers and password managers to correctly identify and fill these fields.
- The community links have been changed to include more options.

I tried to avoid any breaking changes for custom themes, but the following could be an issue:

- ⚠️ ~Automatic redirect to Admin Console~ (this is now controlled by a theme property)
- ⚠️ The `localUser` property is now always present (non-null) in the template, which could break certain conditions.
- ⚠️ The `documentationUrl` and `displayCommunityLinks` properties have been removed.
- ⚠️ We're now loading PatternFly 5 styles only, which might be a breaking change for users extending the `keycloak` theme relying on PatternFly 4.

See the screenshots below for an idea of the different states the UI can be in:

**Initial state (no admin user)**

![Screen Shot 2023-12-13 at 11 51 49](https://github.com/keycloak/keycloak/assets/695720/dedeff22-f6b8-4cf1-bb12-1f4b096c394a)

**Form submitted (with error)**

![Screen Shot 2023-12-13 at 11 54 26](https://github.com/keycloak/keycloak/assets/695720/6e267225-03c0-4bf4-b250-31a8429fe4ee)

**Form submitted (with success)**

![Screen Shot 2023-12-13 at 11 57 19](https://github.com/keycloak/keycloak/assets/695720/c7190821-4d6c-421c-a7ae-dd5f2e406618)

**Local access required**

![Screen Shot 2023-12-13 at 12 09 03](https://github.com/keycloak/keycloak/assets/695720/4daa3d86-311a-4431-be46-1893547b0dac)

**Admin Console disabled, or redirect disabled**

![Screen Shot 2023-12-13 at 12 06 32](https://github.com/keycloak/keycloak/assets/695720/4dff2c50-f281-4898-8313-54217d904a9f)

Closes #21343

jonkoops committed:
Upgrade Welcome theme to PatternFly 5

Closes #21343

Co-authored-by: Pedro Igor <pigor.craveiro@gmail.com>
Signed-off-by: Jon Koops <jonkoops@gmail.com>

jonkoops commented:
@andymunro could I ask you to review the copy in these screenshots? I'd like to make sure the tone of voice and language is consistent with our brand identity.

ssilvert commented:
On User Successfully Created screen, text saying "...you will first have to create an administrative user" should be hidden.

ssilvert commented on themes/src/main/resources/theme/keycloak/welcome/theme.properties:
```diff
@@ -1,7 +1,7 @@
-styles=css/welcome.css
 import=common/keycloak
```
Do we still need this?

ssilvert replied:
I think that any common asset for the Welcome page should be considered a global resource since does not belong to a realm or a theme.  The Welcome `theme type` is an odd duck and is actually global in and of itself.  I suspect its resources should be handled the same way we did the common React library?

It's a little weird that these global resources now live in the legacy `keycloak` theme, which is destined to be deprecated.  If it's easy to fix this in this PR then let's go ahead and do it.  Otherwise, please create an issue for it and add to the issues list for the "commons" epic, https://github.com/keycloak/keycloak/issues/25552.

jonkoops replied:
> I think that any common asset for the Welcome page should be considered a global resource since does not belong to a realm or a theme.

It might not depend on a realm, but it is technically part of a theme, specifically `keycloak`. I am not entirely sure what a 'global resource' would mean here.

> The Welcome theme type is an odd duck and is actually global in and of itself. I suspect its resources should be handled the same way we did the common React library?

Yeah, it's a bit of a strange one, it still always belongs to a theme though, so I think it's fine to keep it's shared resources in the theme itself.

> It's a little weird that these global resources now live in the legacy keycloak theme, which is destined to be deprecated.

I don't think we'll be getting rid of the `keycloak` theme anytime soon, as we still have the `email` and `login` theme types in there. I'd actually propose we merge all the `keycloak.vX` themes back into `keycloak`, as we no longer have code for the original themes around in the long run, just adding confusion.

ssilvert reviewed (commented):
I don't see where PatternFly 3 is actually removed from Keycloak?

ssilvert commented:
This is really good.  Hope we can get it merged soon.

jonkoops commented:
> On User Successfully Created screen, text saying "...you will first have to create an administrative user" should be hidden.

Fixed in 18104de36d4a22b3b8c6dfb0a76cc351d0b4378c, updated the screenshot in the PR description as well.

> I don't see where PatternFly 3 is actually removed from Keycloak?

It's not, it is still used by the login theme. We'll have to get rid of it once we move away from it there.

andymunro reviewed (commented):
Looks good  @jonkoops.  I recommend just two changes.
- Change ".., you will have to first create an administrative user"
  to ".., you first create an administrative user"
- Use initial caps for Administration Console

jonkoops commented:
Updated the copy under 7b3053890534b006ac7c26e22c9884fe6045c4d7, also updated the screenshots with the latest versions. Does it look good to you like this @andymunro?

andymunro reviewed (approved):
Looks good, @jonkoops. Thanks.

xianli123 commented:
Hi @jonkoops  It looks good. One point we can polish is that we can center-align the left form and the right content. What do you think?

hmlnarik commented:
I like it, thank you!

jonkoops commented:
> I'd go one step further and remove the text describing what Keycloak is, and get rid of all the links. Just make it a simple bootstrap the first admin user page with nothing else.

Done under 2a40fd7f79ce560732e73f0279542d102d1b87e4, updated the PR description with the latest screenshots as well. The only caveat with this is that the screen is very empty when the Administration Console is disabled, showing only the Keycloak logo.

jonkoops commented on services/src/main/java/org/keycloak/services/resources/WelcomeResource.java:
```diff
@@ -177,23 +175,30 @@ private Response createWelcomePage(String successMessage, String errorMessage) {
+            URI adminUrl = session.getContext().getUri(UrlType.ADMIN).getBaseUriBuilder().path("/admin/").build();
+
+            // Redirect to the Administration Console if the admin account is already created.
+            if (!bootstrap && adminConsoleEnabled && successMessage == null) {
+                return Response.status(302).location(adminUrl).build();
+            }
```
Yes, this is one of the issues I wanted to discuss (see PR description). I think a property is likely the right way to go here.

jonkoops commented:
I've added a new `redirectToAdmin` property that can be used to customize the redirect behaviour under 27d67f6e4b35f2d7eb4fa321179f7fc34d4795f2, that should prevent custom themes from exhibiting unexpected behaviour.

jonkoops commented:
@stianst can I ask you to give this another pass?

jonkoops commented on themes/src/main/resources/theme/keycloak/welcome/theme.properties:
```diff
@@ -1,7 +1,7 @@
 import=common/keycloak
 
-stylesCommon=node_modules/patternfly/dist/css/patternfly.css node_modules/patternfly/dist/css/patternfly-additions.css
+stylesCommon=node_modules/@patternfly-v5/patternfly/patternfly.min.css node_modules/@patternfly-v5/patternfly/patternfly-addons.css
```
The fact that we're not loading PF5 styles could be a breaking change for users extending the `keycloak` theme.

ssilvert replied:
I think that's OK.

jonkoops commented on themes/src/main/resources/theme/keycloak/welcome/theme.properties:
```diff
@@ -1,7 +1,7 @@
+stylesCommon=node_modules/@patternfly-v5/patternfly/patternfly.min.css node_modules/@patternfly-v5/patternfly/patternfly-addons.css
+styles=css/welcome.css
 
-documentationUrl=https://www.keycloak.org/documentation.html
-displayCommunityLinks=true
```
I've removed these properties, which might be a breaking change for themes extending the default welcome theme.

stianst commented on services/src/main/java/org/keycloak/services/resources/WelcomeResource.java:
```diff
@@ -16,6 +16,7 @@
  */
 package org.keycloak.services.resources;
 
+import jakarta.ws.rs.core.*;
```
Please revert this change to use `*` imports. I know we don't have a proper style guide, but this is one of the things that is good to avoid.

jonkoops replied:
IntelliJ did this for me, so perhaps we should add some configs to the repo to disable that default behaviour. Can you tell me why this is something we try to avoid?

jonkoops replied:
Fixed in 5fac8668ca5ccec016830d5c6ff65b072c8cb117

jonkoops commented:
Managed to get all the tests working, can I ask for another round of reviews? Then let's get this merged so I can work on the release notes :mechanical_arm:

pedroigor commented:
Missing release notes but it seems you are going to update it in a follow-up.
